src/wp-admin/includes/file.php

```diff
@@ -1946,6 +1946,68 @@ function copy_dir( $from, $to, $skip_list = array() ) {
 	return true;
 }
 
+/**
+ * Moves a directory from one location to another via the rename() PHP function.
+ * If the renaming failed, falls back to copy_dir().
+ *
+ * Assumes that WP_Filesystem() has already been called and setup.
+ *
+ * @since 6.2.0
+ *
+ * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
+ *
+ * @param string $from        Source directory.
+ * @param string $to          Destination directory.
+ * @return true|WP_Error True on success, WP_Error on failure.
+ */
+function move_dir( $from, $to ) {
+	global $wp_filesystem;
+
+	$result = false;
+
+	/**
+	 * Fires before move_dir().
+	 *
+	 * @since 6.2.0
+	 */
+	do_action( 'pre_move_dir' );
+
+	if ( 'direct' === $wp_filesystem->method ) {
+		$wp_filesystem->rmdir( $to );
```

Comment:
Would this mean the `$to` dir should be empty? If not empty, should the move/copy fail? (I.e. imagine an error where `$to` is set to `wp-admin` or `wp-content/plugins` for example).

Reply:
From https://www.php.net/manual/en/function.rename.php
> Attempts to rename from to to, moving it between directories if necessary. If renaming a file and to exists, it will be overwritten. If renaming a directory and to exists, this function will emit a warning.

I believe you would only be able to overwrite `wp-admin` or `wp-content/plugins` if the `$from` was also `wp-admin` or `wp-content/plugins`.

Reply:
> I believe you would only be able to overwrite...

Was mostly worried about `$wp_filesystem->rmdir()` but it's called with the default setting of `$recursive = false` so seems the dir will have to be empty to be deleted.

Reply:
Another question here: should "something" happen when `$wp_filesystem->rmdir()` fails, i.e. the `$to` dir is not empty? Currently this is ignored, and PHP's `rename()` would throw a warning. Is that the best behaviour or should it fall back to using `copy_dir()`?

Reply:
`$wp_filesystem->rmdir()` calls `$wp_filesystem->delete()` I've never seen that fail, have you.

The warning should flow through and be returned. I don't think that's necessary but we could put the `rename` in a guard for that so it would drop to `copy_dir`

Reply:
@azaozz what do you think of this update?

Reply:
> could put the rename in a guard for that so it would drop to copy_dir

Looks good imho.